src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/SymbolLookup.java

```diff
@@ -0,0 +1,97 @@
+ * questions.
+ */
+package jdk.incubator.foreign;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.reflect.CallerSensitive;
+import jdk.internal.reflect.Reflection;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A symbol lookup. Exposes a lookup operation for searching symbols, see {@link SymbolLookup#lookup(String)}.
+ * <p> Unless otherwise specified, passing a {@code null} argument, or an array argument containing one or more {@code null}
+ * elements to a method in this class causes a {@link NullPointerException NullPointerException} to be thrown. </p>
+ */
+@FunctionalInterface
+public interface SymbolLookup {
+
+    /**
+     * Looks up a symbol with given name in this lookup.
+     *
+     * @param name the symbol name.
+     * @return the memory address associated with the library symbol (if any).
+     */
+    Optional<MemoryAddress> lookup(String name);
+
+    /**
+     * Obtains a symbol lookup suitable to find symbols in native libraries associated with the caller's classloader
+     * (that is, libraries loaded using {@link System#loadLibrary} or {@link System#load}).
+     * <p>
+     * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
+     * Restricted method are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     *
+     * @return a symbol lookup suitable to find symbols in libraries loaded by given classloader.
+     */
+    @CallerSensitive
+    static SymbolLookup loaderLookup() {
+        Class<?> caller = Reflection.getCallerClass();
+        Reflection.ensureNativeAccess(caller);
+        ClassLoader loader = Objects.requireNonNull(caller.getClassLoader());
+        return loaderLookup0(loader);
+    }
+
+    /**
+     * Obtains a symbol lookup suitable to find symbols in native libraries associated with the given classloader
+     * (that is, libraries loaded using {@link System#loadLibrary} or {@link System#load}).
+     * <p>
+     * This method is <a href="package-summary.html#restricted"><em>restricted</em></a>.
+     * Restricted method are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     *
+     * @param loader the classloader whose symbol lookup is to be retrieved.
+     * @return a symbol lookup suitable to find symbols in libraries loaded by given classloader.
+     */
+    @CallerSensitive
+    static SymbolLookup loaderLookup(ClassLoader loader) {
+        Reflection.ensureNativeAccess(Reflection.getCallerClass());
+        Objects.requireNonNull(loader);
```

Comment:
why do we filter null (bootstrap) loader?

Reply:
Because boot loader is under control of the JDK, not the user. I don't think it's desirable to introduce coupling between user code and set of libraries which happen to be loaded by boot loader.